CMakePresets.json

```
@@ -0,0 +1,76 @@
+{
+    "version": 4,
+    "include": ["cmake/presets/ci.json"],
+    "configurePresets": [
+        {
+            "name": "base",
+            "hidden": true,
+            "binaryDir": "${sourceDir}/build-${presetName}"
+        },
+        {
+            "name": "base-dev",
+            "hidden": true,
+            "cacheVariables": {
+                "CMAKE_BUILD_TYPE": "Debug",
+                "GINKGO_DEVEL_TOOLS": "ON",
+                "CMAKE_EXPORT_COMPILE_COMMANDS": "ON",
+                "CMAKE_COLOR_DIAGNOSTICS": "ON",
+                "CMAKE_LINK_DEPENDS_NO_SHARED": "ON"
+            },
+            "warnings": {
+                "dev": true,
+                "deprecated": true,
+                "uninitialized": true
+            }
+        },
+        {
+            "name": "base-dev-gcc-like",
+            "hidden": true,
+            "generator": "Ninja",
+            "cacheVariables": {
+                "CMAKE_CXX_FLAGS": "$env{CXXFLAGS} -g -fno-omit-frame-pointer -Wpedantic -Wall -Wextra -Wconversion -Wunused"
+            },
+            "condition": {
+                "type": "notEquals",
+                "lhs": "${hostSystemName}",
+                "rhs": "Windows"
+            }
+        },
+        {
+            "name": "release",
+            "inherits": "base",
+            "cacheVariables": {
+                "CMAKE_BUILD_TYPE": "Release"
+            },
+            "warnings": {
+                "dev": false
+            }
+        },
+        {
+            "name": "develop-gcc",
+            "inherits": ["base", "base-dev", "base-dev-gcc-like"],
+            "cacheVariables": {
+                "CMAKE_CXX_COMPILER": "g++"
+            }
+        },
+        {
+            "name": "develop-clang",
+            "inherits": ["base", "base-dev", "base-dev-gcc-like"],
+            "cacheVariables": {
+                "CMAKE_CXX_COMPILER": "clang++"
+            }
+        },
+        {
+            "name": "develop-msvc",
+            "inherits": ["base", "base-dev"],
+            "cacheVariables": {
+                "CMAKE_CXX_FLAGS": "$env{CXXFLAGS} /permissive- /Oy-"
+            },
+            "condition": {
+                "type": "equals",
+                "lhs": "${hostSystemName}",
+                "rhs": "Windows"
+            }
+        }
+    ]
```

Would it makes sense to also add `buildPresets` like
```suggestion
    ], 
      "buildPresets": [
    {
      "name": "develop-gcc",
      "configurePreset": "develop-gcc"
    }
    ]
```
that way one can build with `cmake --build --preset develop-gcc` without knowing the exact `binaryDir`

---

For the 'normal' configuration I can see this making sense. But for the CI configs it would probably be too much.

---

I think `buildPresets` make especially sense in combination with the [workflow]( https://cmake.org/cmake/help/latest/manual/cmake-presets.7.html#workflow-preset) option. So that a user can configure and build a preset with a single command. However, workflows are only available for version 6 and above, which needs cmake 3.25.

---

Just another thought on the buildPresets, we could also add a format preset  with a targets
 key
 ```
"targets": [
        "format",
      ]
```
However, I don't know how to then just pick the "current" `binaryDir` via the preset.

---

I think we'll probably move away from that with pre-commit handling the formatting entirely?

---

The thing I like about presets is that it allows you to run `cmake --build --preset format` from the source directory without knowing the corresponding build directory. However, you are right that moving towards precommit makes it unnecessary.